### PR TITLE
Add CI check to detect Thrift changes

### LINKED ISSUE #3095: Add GitHub build action that fails the build if Thrift generated code is different
Add a build step that uses Thrift to generate the java code and compare it to the code in core/src/main/thrift-gen-java/. If different, fail the build so that the PR owner can regenerate the Thrift java files and push them to their branch. This will require installing Thrift as part of the GitHub action. We should look to see if there is an ASF repository where we could host a Thrift binary tarball and then download it as part of the action.

### PR DESCRIPTION
**Pull request by ctubbsii**
Add a GitHub Actions check that re-generates the Thrift generated code, and compares the alterations to what's committed to the repo. The check will fail if Thrift generates something that isn't checked in. This ensures that we can ignore generated code changes when reviewing pull requests, because if the generated code contained anything unexpected, this check would make it obvious.

This fixes #3095

### CODE REVIEW

**Comment by ctubbsii** [comment]
I'm just glad that the binary I built in Fedora worked on the Ubuntu image... for the first few commits in this PR, I wasn't sure. But it turned out I was just forgetting to `chmod +x` the file. But, I'm glad those first few failed, because the end result is better for having to troubleshoot that.

**Comment by ctubbsii** [comment]
I'm only committing this to main, because I don't expect any thrift changes in earlier branches. But it works the same across all branches... just need to upload a different version of thrift for the different branches. This is actually a variation on what I use locally to resolve the different built versions of Thrift for the different maintenance branches.

If anybody is curious, that script is:

```bash
#! /usr/bin/bash

thriftroot=$HOME/apps/thrift
set -e
thriftdefault="0.16.0"
rootDir=$(git rev-parse --show-toplevel 2>/dev/null ) || ver=$thriftdefault
ver=$({ xmllint --shell "$rootDir/pom.xml" <<<'xpath /*[local-name()="project"]/*[local-name()="properties"]/*[local-name()="thrift.version"]/text()' | grep content= | cut -f2 -d=; } 2>/dev/null || echo "$thriftdefault")
ver=${ver%%-*}

thrift=$thriftroot/thrift-$ver/compiler/cpp/thrift
[[ ! -x $thrift ]] && thrift=$thriftroot/thrift-$thriftdefault/compiler/cpp/thrift

$thrift "$@"
```

I just put this in my local PATH (in `~/bin/thrift`) and I extract copies of thrift releases into `~/apps/thrift` in separate directories. And I build them there, until they fail (they usually fail), but by then the thrift code generator is compiled... and that's all you really need.

**Review by EdColeman — APPROVED** [review]

File: contrib/ci/install-thrift.sh [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+# Download the correct version of Thrift to do CI checks
+
+set -e
+
+thriftdefault="0.16.0"
+rootDir=$(git rev-parse --show-toplevel 2>/dev/null) || ver=$thriftdefault
+ver=$({ xmllint --shell "$rootDir/pom.xml" <<<'xpath /*[local-name()="project"]/*[local-name()="properties"]/*[local-name()="thrift.version"]/text()' | grep content= | cut -f2 -d=; } 2>/dev/null || echo "$thriftdefault")
+ver=${ver%%-*}
+
+sudo wget "https://dist.apache.org/repos/dist/dev/accumulo/devtools/thrift-$ver/thrift" -O /usr/local/bin/thrift &&
```
[the remark this answers is not in the document]
Comment (ctubbsii) [reply contrib/ci/install-thrift.sh:30]: I just created it and uploaded a thrift binary that I made myself. We have a dist area for releases, and one for development. Lots of projects use the one for development for staging release candidates. We have a different process for that, so we've never used this area for anything. I found a good use for it for this.

**Comment by ctubbsii** [comment]
Backported to 2.1.2 as b59dfa9564da59a3e7dff79dd0220aae45af25bf

**Pull request merged**